src/learning/nnet.rs

```diff
@@ -343,24 +325,17 @@ impl<'a, T: Criterion> BaseNeuralNet<'a, T> {
                 g += zeros.vcat(&self.criterion.reg_cost_grad(non_bias_weights));
             }
 
-            capacity += g.cols() * g.rows();
-            grad.push(g / (inputs.rows() as f64));
-        }
-
-        let mut gradients = Vec::with_capacity(capacity);
-
-        for g in grad {
-            gradients.append(&mut g.into_vec());
+            gradients.append(&mut (g / inputs.rows() as f64).into_vec());
         }
 
         // Compute the cost
         let mut cost = self.criterion.cost(&activations[activations.len() - 1], targets);
 
         // Add the regularized cost
         if self.criterion.is_regularized() {
-            for i in 0..self.layer_sizes.len() - 1 {
-                cost += self.criterion.reg_cost(self.get_non_bias_weights(weights, i));
-            }
+            cost += (0..self.layer_sizes.len() - 1)
+                .map(|i| self.criterion.reg_cost(self.get_non_bias_weights(weights, i)))
+                .sum();
```

Comment:
The `sum` method is unstable which is causing the travis build to fail. If using an iterator like this gives significant gains in performance then you can use `fold` instead? Though as the function within the loop is fairly intense I'm not sure it would make much difference.